### PR TITLE
custom mapping changes commit

### PR DESCRIPTION
**Pull request by DhaaraniCIT**

### CODE REVIEW

File: src/app/core/services/misc/mapping.service.ts [hunk 1]
```diff
@@ -87,4 +90,42 @@ export class MappingService {
     return this.apiService.post(`/workspaces/${this.workspaceId}/mappings/settings/`, mappingSettingPayload);
   }
 
+  getMappingStats(sourceType: string, destinationType: string): Observable<MappingStats> {
+    return this.apiService.get(`/workspaces/${this.workspaceId}/mappings/stats/`, { source_type: sourceType, destination_type: destinationType });
+  }
+
+  postMapping(mapping: MappingPost) {
```
Comment (ashwin1111) [inline src/app/core/services/misc/mapping.service.ts:97]: return type

File: src/app/core/services/misc/mapping.service.ts [hunk 2]
```diff
@@ -87,4 +90,42 @@ export class MappingService {
...
+
+  deleteMappingSetting(mappingSettingId: number): Observable<{}> {
+    return this.apiService.delete(`/workspaces/${this.workspaceId}/mappings/settings/${mappingSettingId}/`);
+  }
+
+  triggerAutoMapEmployees(): Observable<{}> {
+    return this.apiService.post(`/workspaces/${this.workspaceId}/mappings/auto_map_employees/trigger/`, {});
+  }
+
+  getXeroField(): Observable<[]> {
```
Comment (ashwin1111) [inline src/app/core/services/misc/mapping.service.ts:128]: `getXeroFields`, return type is incorrect

File: src/app/integration/main/mapping/custom-mapping/custom-mapping.component.ts [hunk 3]
```diff
@@ -201,20 +200,16 @@ export class CustomMappingComponent implements OnInit {
...
       return !this.mappingSettings.some(mapping => mapping.source_field === field.attribute_type);
     });
-    this.qboFields = [MappingDestinationField.CLASS, MappingDestinationField.DEPARTMENT, MappingDestinationField.CUSTOMER];
-
-    const existingQBOFields = this.mappingSettings.filter((mappingSetting: MappingSetting) => {
-      return (mappingSetting.destination_field === MappingDestinationField.CLASS || mappingSetting.destination_field === MappingDestinationField.DEPARTMENT || mappingSetting.destination_field === MappingDestinationField.CUSTOMER);
-    }).map((mappingSetting: MappingSetting) => mappingSetting.destination_field);
-
-    this.qboFields = this.qboFields.filter((qboField: MappingDestinationField) => !existingQBOFields.includes(qboField));
+    this.mappingService.getXeroField().subscribe(
```
Comment (ashwin1111) [inline src/app/integration/main/mapping/custom-mapping/custom-mapping.component.ts:203]: `this.mappingService.getXeroField()` this call can be clubbed along with other API calls done in `setupSettingsAndSetupPage()`

The reason is, they can go in parallel and reduce time. Nesting them will cause them to wait. 1st call succeeds and only after that 2nd request would be sent

File: src/app/integration/main/mapping/custom-mapping/custom-mapping.component.ts [hunk 4]
```diff
@@ -201,20 +200,16 @@ export class CustomMappingComponent implements OnInit {
...
+      result => {
+        this.xeroFields = result;
+        this.xeroFields = this.xeroFields.filter((xeroField: any) => !this.mappingSettings.includes(xeroField));
```
Comment (ashwin1111) [inline src/app/integration/main/mapping/custom-mapping/custom-mapping.component.ts:206]: replace any with proper relevant type

File: src/app/integration/main/mapping/custom-mapping/custom-mapping.component.ts [hunk 5]
```diff
@@ -253,9 +248,12 @@ export class CustomMappingComponent implements OnInit {
       this.mappingService.getMappingSettings(),
       this.mappingService.getFyleExpenseFields()
     ]).subscribe(responses => {
-      this.mappingSettings = responses[0].results;
-      this.fyleFields = responses[1];
-
+      this.mappingSettings = responses[0].results.filter((mappingSetting: MappingSetting) => {
+      return (mappingSetting.destination_field !== MappingDestinationField.ACCOUNT && mappingSetting.destination_field !== MappingDestinationField.BANK_ACCOUNT && mappingSetting.destination_field !== MappingDestinationField.CONTACT && mappingSetting.destination_field !== MappingDestinationField.TAX_CODE) && !mappingSetting.import_to_fyle;
```
Comment (ashwin1111) [inline src/app/integration/main/mapping/custom-mapping/custom-mapping.component.ts:252]: indent is off

File: src/app/integration/main/mapping/mapping-routing.module.ts [hunk 6]
```diff
@@ -2,18 +2,13 @@ import { NgModule } from '@angular/core';
...
 import { MappingComponent } from './mapping.component';
 
 const routes: Routes = [
   {
     path: '',
     component: MappingComponent,
     children: [
-      {
-        path: 'employee',
-        component: EmployeeMappingComponent
```
Comment (ashwin1111) [inline src/app/integration/main/mapping/mapping-routing.module.ts:15]: incase if EmployeeMappingComponent is copied, please delete them

File: src/app/shared/components/mapping/generic-mapping/generic-mapping.component.ts [hunk 7]
```diff
@@ -188,7 +188,7 @@ export class GenericMappingComponent implements OnInit {
     this.sourceType = this.route.snapshot.params.source_field;
     this.mappingService.getMappingSettings().subscribe((mappingSettingResponse: MappingSettingResponse) => {
       const mappingSetting = mappingSettingResponse.results.filter((mappingSetting) => mappingSetting.source_field === this.sourceType.toUpperCase());
-      this.mappingSetting = mappingSetting.length ? mappingSetting[0] : {source_field: FyleField.CATEGORY, destination_field: QBOField.ACCOUNT};
+      this.mappingSetting = mappingSetting.length ? mappingSetting[0] : {source_field: FyleField.CATEGORY, destination_field: FyleField.PROJECT};
```
Comment (ashwin1111) [inline src/app/shared/components/mapping/generic-mapping/generic-mapping.component.ts:191]: `destination_field` should be Xero Account and not Fyle Project

File: src/app/shared/components/mapping/mapping-table/mapping-table.component.ts [hunk 8]
```diff
@@ -42,13 +42,13 @@ export class MappingTableComponent implements OnInit {
       source: searchForm.value.source
     });
 
-    if (this.sourceType?.toUpperCase() === EmployeeFieldMapping.EMPLOYEE) {
-      selectedRow.qbo.id = selectedOption.id;
+    if (this.sourceType?.toUpperCase() === TenantFieldMapping.TENANT) {
```
Comment (ashwin1111) [inline src/app/shared/components/mapping/mapping-table/mapping-table.component.ts:45]: this condition is not required, this is specific to QBO

**Review by ashwin1111 — CHANGES_REQUESTED** [review]

File: src/app/core/services/misc/mapping.service.ts (hunk 1 above)
Comment (DhaaraniCIT) [reply src/app/core/services/misc/mapping.service.ts:97]: what type I should assign for this, in QBO it is employeeMapping, but in xero we don't have that

Comment (ashwin1111) [reply src/app/core/services/misc/mapping.service.ts:97]: Please check the API response and create a type if it doesn't exist

**Review by ashwin1111 — APPROVED** [review]

**Pull request merged**